### ISSUE TITLE
Support fine grained module level build definitions

### ISSUE THREAD

**Issue by aherrmann** [tweag/rules_haskell#1552]
# Goal

Add a rule `haskell_module` to support fine grained build definitions at the level of individual Haskell modules.

# Motivation

The existing `haskell_library|binary|test` rules operate on the level of [Haskell packages](https://downloads.haskell.org/ghc/latest/docs/html/users_guide/packages.html#packages). In these rules building consist of two main steps with corresponding Bazel build actions: Compile all modules of a package in one build action and link all modules together into a library in a separate action (one for static and one for dynamic linking). Additionally, there is an action to construct the package's package database and several supporting actions.

This means that all modules of such a package need to be recompiled even if only a single module in a package changed compared to a previous cached build.

With a `haskell_module` rule we could instead issue one compile action per Haskell module. This would allow us to benefit from Bazel's caching and parallel builds on the level of individual modules.

# Outline

Not all the details are worked out at this stage, however, the idea is that the `haskell_module` rule may be used like this:

```
haskell_module(
    name = "ModuleA",
    src = "ModuleA.hs",
    deps = ["@stackage//:base"],
)
haskell_module(
    name = "ModuleB",
    src = "ModuleB",
    deps = [
        "@stackage//:base",
        ":ModuleA",
    ],
)
haskell_library(
    name = "my-library",
    modules = [
        ":ModuleA",
        ":ModuleB",
    ],
    deps = ["@stackage//:base"],
)
```
In this example the `haskell_module` rules perform the compilation of modules to object and interface files, and the `haskell_library` rule performs the linking of these compiled modules into a static archive or shared object and the generation of a package database for the package.

# Alternatives considered

The goal is to provide what we called "sub-target incremental builds" in the context of the current `haskell_library|binary|test` rules. I.e. avoid recompiling modules that have not changed between builds even if another module that is part of the same library or executabe target did change. An alternative approach that we considered is to use [Bazel's persistent worker](https://www.tweag.io/blog/2019-09-25-bazel-ghc-persistent-worker-internship/) mechanism to provide a persistent GHC session that caches previous compilation outputs (object and interface files). Downsides of this approach are that:
- It is a more complex solution, since it requires maintaining our own cache and maintaining state across worker invocations.
- It introduces potential inhermeticity, since the worker would persist state across build actions outside of Bazel's control.
- The benefits are limited to a worker instance, i.e. modules cannot be shared across build nodes through a remote cache.

# Anticipated problems

## Slower uncached builds

Starting one GHC session per module may well be slower than starting a single GHC session that compiles multiple modules at once. Meaning there is a tradeoff between uncached, builds from scratch and incremental, cached builds. For everyday development the incremental, cached build is likely to be the more common case than uncached, builds from scratch. Meaning, the benefit may outweigh in the common case. Furthermore, we could still use a persistent worker to save on the repeated start-up of GHC sessions similar to how it is already done for Java or Scala in Bazel. In this case we would not need to maintain a cache of compilation outputs in the persistent worker, reducing the downside of this approach.

`rules_haskell` already contains [an implementation of a persistent worker](https://github.com/tweag/rules_haskell/pull/1055) that could be adapted to this use-case.

## More verbose build definitions

Haskell compilation has to occur in dependency order: The Haskell compiler requires interface files of modules that the current module depends on as an input in order to compile the current module. This means that `haskell_module` rules need to spell out the Haskell module dependency graph in the BUILD files. Manually maintaining this dependency graph in both the Haskell sources and the BUILD files is tedious. BUILD file generation, e.g. a [Gazelle extension](https://github.com/bazelbuild/bazel-gazelle/blob/d038863ba2e096792c6bb6afca31f6514f1aeecd/extend.rst) for Haskell, can solve this issue.

**Comment by facundominguez** [comment]
> Another option is to modify what we currently have slightly by having the haskell_module() rule detect when it is being built without a package name supplied (typically from being used by a haskell_library()) and not build anything in that case

Do `haskell_binary` rules supply a package name to `haskell_module`s?

**Comment by facundominguez** [comment]
At first, I expected the abstract approach to help avoid setting `ghcopts` for every module of a library. But now I realize that perhaps this can be done with the explicit approach via transitions. If that is that case, then this becomes pretty much an implementation choice. I don't think it would be obvious for the user to observe the difference, unless she goes into peering at the actions that implement the rules.

I would expect the abstract approach to be simpler, because all the information is gathered in `haskell_module` to later build in the library or binary rules without using transitions. Transitions, in my opinion, spread the task among more concepts than we would otherwise need.

**Comment by facundominguez** [comment]
I don't see much benefit in the explicit API. If a module needs to be compiled with specific `ghcopts`, why not use the `OPTIONS_GHC` pragma instead? Are there any other attributes that would be desirable to specify per module?

**Comment by facundominguez** [comment]
I'm currently trying to implement 3, however, it feels like going against the current so far.

The task is creating the compile actions for the modules, given the descriptions of the haskell_module rules that have been deferred.

My first try was making a recursive function, for every `haskell_module` target, the function would create first the actions to compile its dependencies, and then it would create the action for the `haskell_module` target.

The above is very simple, but it doesn't work because bazel starlark doesn't allow recursive functions!

My next attempt was to encode the recursion with an explicit stack. This seems to work, but since starlark doesn't offer `while` loops, I resorted to a `for` loop over a very long range, that is terminated early when the stack-encoded recursion finishes.

Is this the right way to go?

Here's the function: https://github.com/tweag/rules_haskell/blob/a7d412ea0cd5ff051bbdacb86595e42d4f8bc004/haskell/experimental/private/module.bzl#L236-L298

**Comment by facundominguez** [comment]
And here's another question: if a module is used in two `haskell_binary` rules, and each binary rule creates its own action to compile the module, we should expect the module to be compiled twice and there is little we can do to avoid this, right?

This is something that affects erasing library boundaries when defining `haskell_module` rules. Before, if modules coming originally from libraries ended up used in multiple binaries, they wouldn't be rebuilt. But this doesn't appear to be the case anymore with deferred actions.

**Comment by aherrmann** [comment]
> My first try was making a recursive function, for every `haskell_module` target, the function would create first the actions to compile its dependencies, and then it would create the action for the `haskell_module` target.
> The above is very simple, but it doesn't work because bazel starlark doesn't allow recursive functions!

Correct, Starlark does not allow recursion, or unbounded loops in general. That said, I don't think this is required in this case.

One difficulty seems to be to calculate the set of transitive module dependencies in `haskell_library|...` rule implementation. Fortunately, you don't have to do this in that place. Instead you can let Bazel do this for you on the level of the `haskell_module` rule by returning a field for transitive dependencies in the `HaskellModuleInfo` provider.

Another important feature of Bazel is that the build actions don't need to be emitted in any particular order. So, you could do something along the lines of this (just a sketch):

```
# Declare module outputs
module_outputs = {
    mod.label: struct(
        interface = ctx.actions.declare_file(...),
        object = ctx.actions.declare_file(...),
    )
    for mod in modules
}

# Declare build actions
for mod in modules:
    outputs = module_outputs[mod.label]
    inputs = []
    for dep in mod.transitive_module_dependencies:
        inputs.append(module_outputs[dep.label].interface)
        inputs.append(module_outputs[dep.label].object)
    _run_ghc(
        outputs = outputs,
        inputs = inputs,
        ...
    )
```

> And here's another question: if a module is used in two `haskell_binary` rules, and each binary rule creates its own action to compile the module, we should expect the module to be compiled twice and there is little we can do to avoid this, right?

Yes, that is true. Given the `odir`, `hidir` constraint, i.e. that all modules of a library or binary need to have their interfaces and objects in the same directory, I don't see much that we can do about this.

**Comment by facundominguez** [comment]
When trying to decide what data structure to use to collect transitive dependencies of `haskell_module`, I'm not finding an obvious choice.

This is `mod.transitive_module_dependencies` in Andreas's sketch. Semantically it would be a set of `Target`s. Bazel recommends using `depset` for collecting values transitively, but I imagine that checking for `Target` equality is overkill, since comparing only the labels would be necessary.

But the other alternatives also seem to impose some merging overhead as explained in the documentation. https://docs.bazel.build/versions/main/skylark/depsets.html

Looks like the most efficient data structure would be `depmap`, something like a depset but dedups by looking at a key instead of looking at the values.

Any thought's on how to proceed here?

**Comment by aherrmann** [comment]
A `depset` seems like the best data structure and I'm not aware of a `depset` of `Target`s being a particular performance concern. As far as I know `depset` doesn't use value equality, but compares hashes instead. I'm not sure how exactly this works in the case of `depset`s of `Target`s off the top of my head, but, there are cases in the Bazel code base where the hash code of a target is calculated based only on its label and configuration key and also avoids frequent re-computation of the hash by caching it. In other words, I'd expect it to behave similarly to the `depmap` you describe.

The more concerning part, performance-wise, is that we need to flatten the `depset` inside `haskell_library|...` to generate all the build actions. However, I don't really see a way around this, and so long as there is no overlap of modules between `haskell_library|...` targets it also wouldn't be quadratic.

**Comment by facundominguez** [comment]
> The more concerning part, performance-wise, is that we need to flatten the depset inside haskell_library|... to generate all the build actions. However, I don't really see a way around this, and so long as there is no overlap of modules between haskell_library|... targets it also wouldn't be quadratic.

We have to flatten the transitive dependencies once per compile action in order to produce the list of input interface files. This does look quadratic on the amount of modules in a library. Are we better in this respect with the stack-encoded recursion?

**Comment by aherrmann** [comment]
> We have to flatten the transitive dependencies once per compile action for a module in order to produce the list of input interface files. This does look quadratic on the amount of modules in a library.

Ah, yes, you're right. I was looking at it the wrong way.

This reminds me that we have a somewhat similar situation with the Haskell toolchain libraries, implemented [here](https://github.com/tweag/rules_haskell/blob/9efbbbecbc37c7c2fb72058f018da90a16621417/haskell/private/haskell_impl.bzl#L687-L692). In that case we exploit the fact that a `depset` can give us the dependencies in a well defined order.

I think something like this could be used here as well. If we can iterate over the modules in the right order, such that A appears before B if B depends on A, then we can avoid iterating over each module's transitive dependencies. Another sketch illustrates what I have in mind:

```
# Declare module outputs
module_outputs = {
    mod: [
        ctx.actions.declare_file(...),  # interface
        ctx.actions.declare_file(...),  # object
    ]
    for mod in modules
}

# List of modules in post-order
# Meaning, if module B depends on module A, then A will appear before B.
ordered_modules = depset(transitive = [
    mod[HaskellModuleInfo].transitive_module_deps  # Created with order = "postorder"
    for mod in modules
])

# Collect module build inputs and declare build actions
# This is the only place where we flatten the depset
module_inputs = {}
for mod in ordered_modules.to_list():
    outputs = module_outputs[mod]
    inputs = depset(
        direct = [
            out
            for dep in mod[HaskellModuleInfo].direct_module_deps
            for out in module_outputs[dep]
        ],
        transitive = [
            module_inputs[dep]  # Will be set by a previous iteration, since all deps were visited before.
            for dep in mod[HaskellModuleInfo].direct_module_deps
        ],
    )
    module_inputs[mod] = inputs
    _run_ghc(
        outputs = outputs,
        inputs = inputs,
        ...
    )
```

This flattens the `depset` of all modules only once, and for each module only iterates over its direct dependencies.

**Comment by facundominguez** [comment]
I'm impressed. That's a clever way to copy the structure of a depset.

However, I'm failing to see the essential difference with the previous sketch. Your last sketch still flattens the transitive inputs once per action, only that it is deferred until the very moment the action is executed. Is not it still quadratic?

Which also brings to the front the fact that once we require transitive inputs for all actions, there is no approach to implement it that isn't potentially quadratic, not even running the actions in the `haskell_module` rules as before.

**Comment by aherrmann** [comment]
> However, I'm failing to see the essential difference with the previous sketch. Your last sketch still flattens the transitive inputs once per action, only that it is deferred until the very moment the action is executed. Is not it still quadratic?

The difference is that it benefits from the sharing of `depset` nodes and avoids unnecessary flattening in the analysis phase. The [performance documentation](https://docs.bazel.build/versions/main/skylark/depsets.html#performance) raises the concern about calling `.to_list()` in rule implementations, not about using `depset`s in general. See also [here](https://docs.bazel.build/versions/main/skylark/performance.html#transitive-action-inputs-should-be-depsets) and [here](https://docs.bazel.build/versions/main/skylark/performance.html#use-ctxactionsargs-for-command-lines).

> Which also brings to the front the fact that once we require transitive inputs for all actions, there is no approach to implement it that isn't potentially quadratic, not even running the actions in the `haskell_module` rules as before.

Yes, if we need those inputs, then we can't avoid that. But, by using `depset` efficiently in the rule implementation we can make best use of Bazel's mechanisms to optimize these kinds of situations.

**Comment by facundominguez** [comment]
Should dependencies be inherited from libraries and binaries?
`gazelle_haskell_modules` copies those into the dependencies of `haskell_module`, but maybe we should leave room to the user to tweak that manually (?).

**Comment by aherrmann** [comment]
> Should dependencies be inherited from libraries and binaries? `gazelle_haskell_modules` copies those into the dependencies of `haskell_module`, but maybe we should leave room to the user to tweak that manually (?).

Good question. I can see two ideals there:
1. A `haskell_module` has no library `deps` but only module `deps`. At compilation we just forward all library deps that are defined at `haskell_library|binary|test` to the module compilation action.
2. Each `haskell_module` has minimal library deps, i.e. only those `haskell(_cabal)_library` that that module actually depends on. At compilation we forward only the module's library deps (and I assume we also need transitive library deps from libraries and modules for template Haskell).

Option 1. is not optimal for incremental builds and parallelism, because a module may be rebuilt when an unused library dependency changes or its build may need to wait for an unused library to finish building. The `unused_inputs_file` feature could potentially help with this, but it's additional complexity to get this to work.

Option 2. is better for incremental builds and parallelism. But, it's harder to get right with `gazelle_haskell_modules`, because we would need a module index for all library targets, including external ones (`@stackage`) to get it right, which Gazelle doesn't have right now.

I don't see an obvious winner between the two. I would err on the side of option 2 and consider option 1 if we know that we can make `unused_inputs_file` work reasonably in this case.

**Comment by facundominguez** [comment]
How about merging `deps` from libraries and the module, and offering a boolean attribute to disable the merging?

I think option (1) could be preferable if we can make `unused_inputs_file` work. So maybe we can just wait until we earn clarity on that before committing to a choice.

**Comment by aherrmann** [comment]
> I think option (1) could be preferable if we can make `unused_inputs_file` work. So maybe we can just wait until we earn clarity on that before committing to a choice.

Agreed, let's do that. 

> How about:
> 3. merging `deps` from libraries and the module, and offering a boolean attribute to disable the merging?

It's essentially letting the user switch between option 1 or 2. Instead of solving the problems of 1 with `unused_inputs_file` it instead defers to the user to fix such issues on demand. I fear that, without a good way for users to generate minimal module deps, it would boil down to option 1 but without `unused_inputs_file` in most use-cases.

**Comment by aherrmann** [comment]
I had an idea regarding cross package module dependencies that came to my mind while looking at #1623. I'm not sure if this is a good idea, but I thought I'll put it out here for consideration.

In this approach `haskell_module` `deps` always express dependencies on modules, if a `haskell_module` depends on a `haskell_library` then that means that it depends on all modules of that library. For dependencies on modules of a different package the `haskell_library|...` needs to depend on the corresponding `haskell_library` target, this is how we can provide the required package-db. Each module dependency of the current module must either be part of the same `haskell_library|...`'s `modules` attribute, or must be provided by a `haskell_library` that is listed in `deps` of the `haskell_library|...` containing the current module. The following snippet illustrates this:
```
haskell_module(
    name = "Foo.A",
    deps = [
        ":Foo.B",  # depends on module Foo.B in the same package
        ":Bar.C",  # depends on module Bar.C in package bar
        ":baz",  # depends on all modules (Baz.D, Baz.E) in package baz
        "@stackage//:lens",  # depends on all modules in package lens
    ],
    ...
)
haskell_module(name = "Foo.B", ...)
haskell_module(name = "Bar.B", ...)

haskell_library(
    name = "foo",
    modules = [":Foo.A", "Foo.B"],
    deps = [":bar", ":baz", "@stackage//:lens"],  # direct deps of package foo - this is how we can know that Bar.C is provided by package bar
    ...
)
haskell_library(
    name = "bar",
    modules = [":Bar.C"],
    deps = ...,
    ...
)
haskell_library(
    name = "baz",
    modules = [":Baz.D", "Baz.E"],
    deps = ...,
    ...
)
```
Let me know what you think.

**Comment by facundominguez** [comment]
Could work. One aspect that I'd like us to figure out is what is the story when using template haskell. Right now a module which uses TH can't be built until the libraries that need to be loaded are built. This means that compiling the module must wait on entire libraries to be built and linked. Could we only load the object files of the imported modules (and their transitive module dependencies)  instead of libraries in these cases?

**Comment by facundominguez** [comment]
We solved the discussion of the abstract vs explicit interface, but it looks like this issue is to stay open until we have implemented all aspects of `haskell_module`.

**Comment by facundominguez** [comment]
A recap of what else needs to be done to complete this issue would be helpful.

**Comment by aherrmann** [comment]
> A recap of what else needs to be done to complete this issue would be helpful.

Agreed, I think in terms of defining and building modules we've covered all that comes to mind:
- [x] define module level dependency graph
- [x] module level incremental builds
- [x] cross package module dependencies
- [x] dependencies on toolchain libraries, Cabal libraries, and regular Haskell libraries
- [x] TH support
- [x] profiling support

There are some things that may work already, but we should have tests for to be sure:
- [ ] GHC plugin support
- [ ] `.hsc` support
- [ ] `.hs-boot` support
- [ ] `c2hs` support

And there are integrations with other rules_haskell features:
- [ ] `haskell_doc` support
- [ ] `haskell_repl` support (including hie-bios output)
- [ ] `haskell_doctest` support

@facundominguez Let me know if something else comes to mind.

**Comment by facundominguez** [comment]
`tests/haskell_module/plugin` and `tests/haskell_module/attr_merging` both test plugins, so perhaps that can be ticked.

Additionally, we should cast `tests/java_classpath` with `haskell_module`.
